### PR TITLE
automatically reload config on changes

### LINKED ISSUE #20: add config watcher
Right now users need to restart the daemon every time they tweak their config. This is sad because restarting the daemon means losing all their sessions, and not losing your sessions is the whole point of shpool. We should use the notify crate to pick up changes to the config file.

### PR DESCRIPTION
**Pull request by ethanpailes**
This patch teaches the shpool daemon to automatically pick up changes to the config when a user writes a new value. Right now, people need to restart the daemon every time they make a config change, which is pretty disruptive because it means losing sessions. Now
people can tweak config without losing sessions.

Fixes #20

### CODE REVIEW

**Review by jrmlhermitte — COMMENTED** [review]
Subway review. Need to look at test in more detail so please don't wait on me if you get another review! :)

File: libshpool/src/attach.rs
Comment (ethanpailes) [reply libshpool/src/attach.rs:122]: Yeah I think I just changed the name because I found the shadowing confusing.

File: libshpool/src/config.rs
Comment (ethanpailes) [reply libshpool/src/config.rs:56]: This is the path building API for rust as far as I know. It will respect the platform path separator correctly (not that this matters much since we only support linux at the moment, but we might want to support windows in the future).

File: libshpool/src/config.rs
Comment (ethanpailes) [reply libshpool/src/config.rs:48]: The `Manager` struct kindof already is this. The path isn't explicitly stored, but it is implicitly captured by the watcher member. Since we don't need to explicitly use the path later, I don't think it makes sense to store the data.

File: libshpool/src/config.rs
Comment (ethanpailes) [reply libshpool/src/config.rs:54]: The path isn't getting returned here. This is part of an expression computing the path and then assigning it to the `config_path` variable.

File: libshpool/src/config.rs
Comment (jrmlhermitte) [reply libshpool/src/config.rs:56]: sounds good makes sense!

File: libshpool/src/config.rs
Comment (ethanpailes) [reply libshpool/src/config.rs:95]: Yes, this can panic, but it will only panic on a poisoned mutex (well technically rwlock). In the case of a poisoned mutex, there really isn't anything reasonable you can do to recover and data might be corrupted. It is a pretty rare condition that never really happens in practice. The recommended practice is to  unwrap since unwinding the stack for the current thread is pretty much the only safe thing to do.

Comment (jrmlhermitte) [reply libshpool/src/config.rs:95]: sounds reasonable, thanks for explaining!

File: libshpool/src/config.rs
Comment (jrmlhermitte) [reply libshpool/src/config.rs:49]: sorry, i thought there was an `unwrap` here. I likely misread while on subway. Sounds good, that sounds reasonable. And thanks for the style insight.

File: libshpool/src/config.rs
Comment (jrmlhermitte) [reply libshpool/src/config.rs:54]: What I meant is while we get the path, there is a side effect of updating the `config`, is that correct?
Do you think a side effect while defining a variable may be harder to read? Is there another way we can do this without?
Or is this an accepted practice? (Im okay with it so long as this is something reviewers can easily notice)

File: libshpool/src/config.rs
Comment (jrmlhermitte) [reply libshpool/src/config.rs:48]: if it's a closure that makes sense. no need to expose it further. i agree, sounds good!

File: libshpool/src/config.rs
Comment (ethanpailes) [reply libshpool/src/config.rs:54]: I did it this way intentionally. I could switch it to assigning both as a side effect or having the expression evaluate to a tuple if you want, but I think this is fairly normal rust.

File: libshpool/src/config.rs
Comment (nicolasavru) [inline libshpool/src/config.rs:70]: nit: would be clearer to put this above the config_path assignment since this only uses config, not config_path.

**Review by nicolasavru — APPROVED** [review]

File: libshpool/src/config.rs
Comment (ethanpailes) [reply libshpool/src/config.rs:54]: I'm switching to the tuple assignment based on rasputin's comment below.

File: libshpool/src/config.rs
Comment (ethanpailes) [reply libshpool/src/config.rs:70]: config get's reassigned in that block. I'm switching to a tuple assignment for both config and config_path at the same time to make it clearer.

**Pull request merged**
